drivers/auxiliary/ups.cpp

```diff
@@ -0,0 +1,577 @@
+    ConnectionSettingsNP[0].fill("TIMEOUT", "Timeout (sec)", "%.0f", 1, 30, 1, 5);
+    ConnectionSettingsNP[1].fill("RETRIES", "Max Retries", "%.0f", 1, 10, 1, 3);
+    ConnectionSettingsNP[2].fill("RETRY_DELAY", "Retry Delay (ms)", "%.0f", 100, 5000, 100, 1000);
+    ConnectionSettingsNP[3].fill("RECONNECT_ATTEMPTS", "Max Reconnect Attempts", "%.0f", 0, 10, 1, 3);
+    ConnectionSettingsNP.fill(getDeviceName(), "CONNECTION_SETTINGS", "Connection", CONNECTION_TAB, IP_RW, 60, IPS_IDLE);
+
+    // Setup update period
+    UpdatePeriodNP[0].fill("PERIOD", "Period (s)", "%.1f", 1, 3600, 1, 10);
+    UpdatePeriodNP.fill(getDeviceName(), "UPDATE_PERIOD", "Update", CONNECTION_TAB, IP_RW, 60, IPS_IDLE);
+
+    // Add UPS parameters as weather parameters
+    addParameter("BATTERY_CHARGE", "Battery Charge", 10, 100, 0);
+    addParameter("BATTERY_VOLTAGE", "Battery Voltage", 12, 14, 0);
+    addParameter("INPUT_VOLTAGE", "Input Voltage", 210, 240, 0);
+
+    // Set critical parameters
+    setCriticalParameter("BATTERY_CHARGE");
+
+    // Load config before setting any defaults
+    loadConfig(true);
+
+    addDebugControl();
+
+    return true;
+}
+
+void UPS::ISGetProperties(const char *dev)
+{
+    INDI::Weather::ISGetProperties(dev);
+
+    // Always define these properties
+    defineProperty(ServerAddressTP);
+    defineProperty(UPSNameTP);
+    defineProperty(ConnectionSettingsNP);
+}
+
+bool UPS::Connect()
+{
+    if (ServerAddressTP[0].getText() == nullptr || ServerAddressTP[1].getText() == nullptr)
+    {
+        LOG_ERROR("Server address or port is not set.");
+        return false;
+    }
+
+    // Ensure socket is closed before reconnecting
+    if (socketFd >= 0)
+    {
+        close(socketFd);
+        socketFd = -1;
+    }
+
+    // Create socket
+    socketFd = socket(AF_INET, SOCK_STREAM, 0);
+    if (socketFd < 0)
+    {
+        LOGF_ERROR("Failed to create socket: %s", strerror(errno));
+        return false;
+    }
+
+    // Setup address structure
+    struct sockaddr_in serverAddr;
+    memset(&serverAddr, 0, sizeof(serverAddr));
+    serverAddr.sin_family = AF_INET;
+    serverAddr.sin_port = htons(std::stoi(ServerAddressTP[1].getText()));
+
+    // Convert hostname to IP
+    struct hostent *he = gethostbyname(ServerAddressTP[0].getText());
+    if (he == nullptr)
+    {
+        LOGF_ERROR("Failed to resolve hostname: %s", ServerAddressTP[0].getText());
+        close(socketFd);
+        socketFd = -1;
+        return false;
+    }
+    memcpy(&serverAddr.sin_addr, he->h_addr_list[0], he->h_length);
+
+    // Set timeout for connect
+    struct timeval tv;
+    tv.tv_sec = static_cast<int>(ConnectionSettingsNP[0].getValue());
+    tv.tv_usec = 0;
+    setsockopt(socketFd, SOL_SOCKET, SO_RCVTIMEO, &tv, sizeof(tv));
+    setsockopt(socketFd, SOL_SOCKET, SO_SNDTIMEO, &tv, sizeof(tv));
+
+    // Enable keep-alive
+    int keepalive = 1;
+    setsockopt(socketFd, SOL_SOCKET, SO_KEEPALIVE, &keepalive, sizeof(keepalive));
```

Comment:
is there a reason for using your own implementation instead of using INDI's own connectionTCP?

Reply:
Indeed, no specific reason, i migrated to indi TCP connection